boards: shields: Add support for WIZ550io shield

gramsay0 opened the pull request:
WIZ550io is a W5500 breakout board with an extra PIC12F519 MCU.
The PIC12F519 MCU initialises the W5500 with a unique MAC address after a GPIO reset.
The PIC12F519 requires a 150ms delay after GPIO reset to configure the W5500.
There is an optional "ready" GPIO signal that can be used to reduce that delay.

decsny commented on dts/bindings/ethernet/wiznet,w5500.yaml:
```diff
@@ -23,3 +23,25 @@ properties:
       The reset pin of W5500 is active low.
       If connected directly the MCU pin should be configured
       as active low.
+  hw-reset-delay:
```
seems like this is a software configuration that should not go in DT based on how it is used. maybe add another compatible to the node that indicates if the w5500 is the io variation?

gramsay0 replied:
I had considered adding a compatible node for this variant, I will have another look into that soon

parthitce commented on dts/bindings/ethernet/wiznet,wiz550io.yaml:
```diff
@@ -0,0 +1,48 @@
+        };
+    };
+
+compatible: "wiznet,wiz550io"
```
What's the technical reason to introduce another compat with yaml here? IMO `dts/bindings/ethernet/wiznet,w5500.yaml` is the same with additional property which can is optional.

decsny replied:
According to the commit message, this PR is adding support for a breakout board with a w5500 chip on it, which is not the same thing as the w5500 chip itself. The breakout board also has an MCU which has extra signals and configurations than the w5500 has on its own. I think it makes sense in this case from the perspective of DT to therefore define a new compatible.

Ill say that someone could make the argument that the new compatible should actually be a parent with a child w5500 node, but for the functionality the PR is appearing to add, I think, while accurate, that would be total overkill.

parthitce replied:
NACK! Underlying IC is the same, which is W5500. So no additional bindings file required here IMHO. For the breakout board, it can potentially be a shield as like `boards/shields/arceli_eth_w5500/`

gramsay0 replied:
I agree with @decsny, I think having a new compat makes sense.
It is more logical to a user to just specify "wiznet,wiz550io" and have the correct defaults applied (like hardware reset delay).
It also avoids polluting "wiznet,w5500" with WIZ550io specific configuration.

> Ill say that someone could make the argument that the new compatible should actually be a parent with a child w5500 node, but for the functionality the PR is appearing to add, I think, while accurate, that would be total overkill.

I took inspiration from "nxp,flexcan-fd" which is a specialization of "nxp,flexcan":
https://github.com/zephyrproject-rtos/zephyr/blob/72a17b6d887ef20b3cc08a25d8ea1d622792d57a/dts/bindings/can/nxp%2Cflexcan-fd.yaml#L12
I think this fits well as it is not really a parent (like an SPI controller to an SPI device), it's just an extension/specialization of "wiznet,w5500"

parthitce replied:
Not a CAN expert, but AFAIK CAN-FD and CAN are two different things underlying IP blocks. W.r.t w5500, WIZ550io = W5500 + PIC12F519 with 2 additional capability to W5500,
1. Unique Hardware MAC address
2. Ready signal

So this needs to be in the same w5500 bindings with only `ready-gpio`. Regarding the MAC address, we have 2 options already,
1. local-mac-address
2. zephyr,random-mac-address
when neither presents, this eventually means hardware unique MAC.

parthitce replied:
```
git diff dts/
diff --git a/dts/bindings/ethernet/wiznet,w5500.yaml b/dts/bindings/ethernet/wiznet,w5500.yaml
index 1490f0763b9..a634e85c9a9 100644
--- a/dts/bindings/ethernet/wiznet,w5500.yaml
+++ b/dts/bindings/ethernet/wiznet,w5500.yaml
@@ -23,3 +23,10 @@ properties:
       The reset pin of W5500 is active low.
       If connected directly the MCU pin should be configured
       as active low.
+  ready-gpios:
+    type: phandle-array
+    description: Ready pin.
+
+      The WIZ550io module has an extra GPIO pin that signals
+      when the PIC12F519 has finished configuring the W5500.
+      This is optional to shorten the hardware reset delay.
```
handle the ready-gpio as an optional property. And create a separate shield for `WIZ550io` under `boards/shields`

parthitce replied:
Regarding the MAC address part, I agree that there isn't any reference. DTS bindings should be specific, so I propose to discuss this hardware unique address changes under `dts/bindings/ethernet/ethernet-controller.yaml`.

decsny replied:
> Regarding the MAC address part, I agree that there isn't any reference. DTS bindings should be specific, so I propose to discuss this hardware unique address changes under dts/bindings/ethernet/ethernet-controller.yaml.

Yes, I also don't like the ambiguity, but I will defer to the author on their opinion of if this is scope creep to discuss on this PR.

As for the shield idea, I would also like to hear @gramsay0 's opinon of this. It's not a wrong idea in general, but I will say I don't think the ready-gpio property can go on the w5500 binding if it is not part of the actual w5500 chip.

gramsay0 replied:
> Not a CAN expert, but AFAIK CAN-FD and CAN are two different things underlying IP blocks. W.r.t w5500, WIZ550io = W5500 + PIC12F519 with 2 additional capability to W5500,

I don't understand the distinction you are making?

Yes, a CAN-FD controller is a different hardware IP block, it provides classical CAN + CAN-FD extensions.
WIZ550io is a hardware module, it provides w5500 + PIC12F519 extensions.

In hardware terms: WIZ550io != w5500.

> So this needs to be in the same w5500 bindings with only `ready-gpio`. Regarding the MAC address, we...

I like the idea of using the unique MAC address is neither local nor random options are enabled.
I can see that being done elsewhere already:
https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/ethernet/eth_esp32.c#L150

Not having a configurable `hw-reset-delay` would mean `reset-gpio`/`ready-gpio` would be required for WIZ550io, which seems ok.

> Yes, I also don't like the ambiguity, but I will defer to the author on their opinion of if this is scope creep to discuss on this PR.

I could add a comment to ethernet-controller.yaml if requested, it does seem a little out of scope.

> As for the shield idea, I would also like to hear @gramsay0 's opinon of this

I am not very familiar with shields, they don't seem very portable with things like `arduino_spi`, `mikrobus_spi`, `boosterpack_spi`, `pico_spi` hard coded..

> I will say I don't think the ready-gpio property can go on the w5500 binding if it is not part of the actual w5500 chip.

This is an excellent point.

parthitce reviewed (changes requested):
Thanks for the changes.

gramsay0 commented on drivers/ethernet/eth_w5500.c:
```diff
@@ -567,10 +577,50 @@ static int w5500_init(const struct device *dev)
 			return -EINVAL;
 		}
-		gpio_pin_set_dt(&config->reset, 0);
+		gpio_pin_set_dt(&config->reset, 1);
```
This just makes sure the reset pin is held low for the minimum specified time:

![image](https://github.com/zephyrproject-rtos/zephyr/assets/124088381/5d3383fc-9eed-4ff1-b831-142f466fa954)

The pin is set high again after the delay

gramsay0 commented on drivers/ethernet/eth_w5500.c:
```diff
@@ -567,10 +577,50 @@ static int w5500_init(const struct device *dev)
+		}
 	}
 
+#if !DT_INST_PROP(0, zephyr_random_mac_address) && !NODE_HAS_VALID_MAC_ADDR(DT_DRV_INST(0))
```
The MAC address needs to be read out after a hardware reset and before software reset.

> Therefore, we recommend HW reset instead of SW reset. Nevertheless, if users want to use SW reset, we recommend to save MAC address ...... before SW reset, and writing those information to WIZ550io after SW reset.

https://docs.wiznet.io/Product/ioModule/wiz550io

parthitce replied:
I understand. In that case, we should avoid soft reset IMO. Makes sense to move this to `w5500_set_macaddr` and `else` condition the `w5500_soft_reset` if ready gpio present?

gramsay0 replied:
I am a little hesitant to remove the soft reset, as I'm not sure exactly what registers the PIC12F519 MCU configures in the W5500. Other than the mac address, registers should be configured by the Zephyr driver (or default/reset values).

I tested your suggestion and it does _seem_ to work.

Given the above, can you confirm whether you think this change should be made, and I will update accordingly

parthitce replied:
Understood thanks

raveious commented on dts/bindings/ethernet/wiznet,w5500.yaml:
```diff
@@ -23,3 +23,9 @@ properties:
       The reset pin of W5500 is active low.
       If connected directly the MCU pin should be configured
       as active low.
+  ready-gpios:
```
I'm wondering if this should be in a new device tree binding that inherits from this one. You're sorta imposing this requirement onto everyone even if they aren't using this module which is what I believe to be the more common case.

gramsay0 replied:
@raveious an earlier revision of this PR had that, the maintainer did not agree, so it was removed

parthitce replied:
I understand the differences between this module and underlying W5500 IC itself. Meaning, there is no ready-gpio from W5500. So to address that we can potentially introduce binding consuming the w5500 binding in include.

raveious reviewed (commented):
Closing my review as my issues were addressed previously.

parthitce reviewed (dismissed):
LGTM, please rebase. To add, does it makes sense to add the overlay based on WIZ550IO?

gramsay0 commented:
> LGTM, please rebase. To add, does it makes sense to add the overlay based on WIZ550IO?

I have added a WIZ550io shield, similar to the arceli_eth_w5500 shield

decsny reviewed (changes requested):
need to determine a different solution for the w5500 io board that fits into the zephyr architecture better

gramsay0 commented on drivers/ethernet/eth_w5500.c:
```diff
@@ -567,10 +577,51 @@ static int w5500_init(const struct device *dev)
 
+#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(ready_gpios)
+	if (config->ready.port) {
+		if (!gpio_is_ready_dt(&config->ready)) {
```
@decsny could you help suggest an alternative to this?

decsny commented:
@gramsay0 I'm a bit confused because I looked for this W5500io module on the internet and found these docs: https://docs.wiznet.io/Product/ioModule/W5500-io#pin-description but they don't say anything about a ready pin as far as I can tell, can you clarify the discrepancy?

decsny commented:
nevermind, I just realized there is a w5500io and a w550io.... that's a bit confusing itself

decsny commented:
@gramsay0 

I noticed the RST pin on the w550io board is actually not the same signal as the RST going to the w5500 controller itself. It actually is a signal that goes to the PIC12F519, and a different pin from the PIC12F519 is connected to the w5500. And this signal is even documented differently, having different timing requirements.

I think this breakout board should have a different compatible, adding a ready signal to the w5500 compatible is incorrect. The DT could look something like this

```
&arduino_spi {
 	status = "okay";

	eth_w5500: eth_w5500@0 {
		compatible = "wiznet,w5500";
                ....
               /* do not put the reset gpio property here */
               no-reset;
	};
}

w550io {
		compatible = "wiznet,w550io";
		reset-gpios = <&arduino_header 14 GPIO_ACTIVE_LOW>;  /* D8 */
		ready-gpios = <&arduino_header 13 GPIO_ACTIVE_HIGH>; /* D7 */
                wiznet,w5500 = <&eth_w5500>
}
```

The issue of having board specific code in the file is actually not a huge problem, but I don't like how it's done right now which is just putting the code within some #if and checking if that property is present, which it shouldn't be since it shouldn't be part of that binding. Also, in my opinion, since these reset signals are not the same as each other, and the whole reset sequence really is different, code should reflect that, right now it's a weird hack looking code that can be misleading to future readers and potentially annoying to maintain. 

To handle the different reset sequence, I would put a boolean DT property in the w5500 binding called "no-reset". I have done some research to confirm that properties like this are common in linux DT bindings to clarify that a device may be reset by another mechanism which is the case here, as the mechanism doing the reset will be the PIC12F519 rather than the host zephyr mcu. If this property is set, don't do the reset sequence in the w5500 init. Instead, you can have a SYS_INIT function in the same file that runs for every w550 node with higher priority than the w5500 init (ie runs before) that will do the appropriate reset sequence for this w550 module involving the longer wait after RST asserted and checking the RDY pin etc. You can also put in this function the spi read to read the mac address and I think you should have no problem setting the data in the w5500_runtime struct from here for the corresponding w5500 device in the phandle.

This way the code for the w550io is separate from the code for the w5500 and the DT bindings are accurate.

gramsay0 commented:
@decsny I agree and really like the approach you have proposed.
I'll have a look into implementing these changes, thanks

gramsay0 commented:
@decsny I have made the modifications you suggested.

Although I'm not certain the WIZ550io code is in the correct location, as it's not _really_ an Ethernet controller. This could be moved elsewhere (drivers/misc?) if you prefer.

Rather than adding a "no-reset" property to W5500 I put a static assert in the WIZ550io driver to ensure the "reset-gpios" property is not set in the W5500 device

decsny commented:
won't the init sequence in the eth_w5500 have some problems now?

gramsay0 commented:
> won't the init sequence in the eth_w5500 have some problems now?

@decsny can you be more specific, I'm not sure what you are referring to?

decsny reviewed (dismissed):
approving because I'm assuming it works but what I meant is, doesn't the w5500 driver init still happen and will it conflict with what the w550io init is doing

gramsay0 commented:
> approving because I'm assuming it works but what I meant is, doesn't the w5500 driver init still happen and will it conflict with what the w550io init is doing

Both the wiz550io and w5500 init occur. wiz550io has two asserts to prevent conflicts:

- wiz550io init occurs first
- w5500 device does not have GPIO reset enabled (SW reset is ok)

If neither "zephyr,random-mac-address" or "local-mac-address" are configured, w5500 leaves the mac address untouched, which in this case will use the wiz550io embedded MAC address

parthitce reviewed (changes requested):
IMO the core view of the PR is fully changed. This is how I understand it as a whole,
- WIZ550io uses W5500 + X IC to configure the MAC to be unique
- This is notified to the consumer over a GPIO line to inform that the W5500 chip ready with unique mac address

So the only difference between W5500 and WIZ550io is to have unique MAC address. Regarding the bindings part, I ACK'd the ready-gpio part, which is arguably in WIZ550io binding (which uses w5500 binding). But technically this is MAC address handling for W5500, which shall reside in w5500 driver itself.

But I don't really understand the reason behind moving ethernet code to `drivers/misc`? Also additional Kconfig? Really IMO this is off track and unmanageable AFAIK. NACK!

MaureenHelm commented:
@parthitce @decsny are you both available to join the dev-review meeting tomorrow to discuss?

parthitce commented:
> @parthitce @decsny @gramsay0 are you available to join the dev-review meeting tomorrow to discuss?

@MaureenHelm Yes, please share the time and method to join.

MaureenHelm commented:
> > @parthitce @decsny @gramsay0 are you available to join the dev-review meeting tomorrow to discuss?
> 
> @MaureenHelm Yes, please share the time and method to join.

https://github.com/zephyrproject-rtos/zephyr/wiki/Zephyr-Committee-and-Working-Groups#zephyr-dev-meeting

parthitce commented:
> > > @parthitce @decsny @gramsay0 are you available to join the dev-review meeting tomorrow to discuss?
> > 
> > 
> > @MaureenHelm Yes, please share the time and method to join.
> 
> https://github.com/zephyrproject-rtos/zephyr/wiki/Zephyr-Committee-and-Working-Groups#zephyr-dev-meeting

I was occupied with another meeting and couldn't come in time. If you have already discussed, happy to sync with the status or otherwise in next call. Thanks

decsny commented:
> > > > @parthitce @decsny @gramsay0 are you available to join the dev-review meeting tomorrow to discuss?
> > > 
> > > 
> > > @MaureenHelm Yes, please share the time and method to join.
> > 
> > https://github.com/zephyrproject-rtos/zephyr/wiki/Zephyr-Committee-and-Working-Groups#zephyr-dev-meeting
> 
> I was occupied with another meeting and couldn't come in time. If you have already discussed, happy to sync with the status or otherwise in next call. Thanks

We discussed about putting the code in the shield board folder instead of drivers/misc which is fine by me but not sure if there is actually infrastructure in place to pull in c code for a shield.... since I didn't find any other shield with c code and didn't see any infrastructure for it. If it is possible then I definitely prefer it over drivers/misc because it is more obviously associated with the board rather than being somewhere else

ubieda commented:
> We discussed about putting the code in the shield board folder instead of drivers/misc which is fine by me but not sure if there is actually infrastructure in place to pull in c code for a shield.... since I didn't find any other shield with c code and didn't see any infrastructure for it. If it is possible then I definitely prefer it over drivers/misc because it is more obviously associated with the board rather than being somewhere else

Following up here: I think you're right. There appears to be no infrastructure for this at the moment.

decsny commented:
> > We discussed about putting the code in the shield board folder instead of drivers/misc which is fine by me but not sure if there is actually infrastructure in place to pull in c code for a shield.... since I didn't find any other shield with c code and didn't see any infrastructure for it. If it is possible then I definitely prefer it over drivers/misc because it is more obviously associated with the board rather than being somewhere else
> 
> Following up here: I think you're right. There appears to be no infrastructure for this at the moment. I initially thought it would work the same way as with boards.

okay, then maybe drivers/misc appears still the most appropriate place we have thought of for it that doesn't require effort outside the scope of the PR

gramsay0 commented:
The dev-review meeting timezone does not work well for me. I agree with the idea that was discussed.

I created a PR to allow adding source code for shields: https://github.com/zephyrproject-rtos/zephyr/pull/75310

gramsay0 committed:
boards: shields: Add support for WIZ550io shield

WIZ550io is a W5500 breakout board with an extra PIC12F519 MCU.
The PIC12F519 MCU initialises the W5500 with a unique MAC address
after a GPIO reset. The PIC12F519 requires a 150ms delay after GPIO
reset to configure the W5500. There is an additional "ready" GPIO
signal that can be used to reduce that delay.

Signed-off-by: Grant Ramsay <gramsay@enphaseenergy.com>

decsny commented:
@parthitce Are you okay with this new approach now? It has gone to extensive review in the dev review meeting and architecture working group, so hopefully this PR can be close to merging after a long time now

decsny commented on boards/shields/wiznet_wiz550io/dts/bindings/wiznet,wiz550io.yaml:
@gmarull @kartben If the binding is in the shield board folder like this, would it be included in the binding documentation generation?

parthitce commented on boards/shields/wiznet_wiz550io/wiznet_wiz550io.c:
```diff
@@ -0,0 +1,136 @@
+#define W5500_NODE(inst) DT_INST_PHANDLE(inst, wiznet_w5500)
+
+#define WIZ550IO_INIT(inst)							\
+	BUILD_ASSERT(!DT_NODE_HAS_PROP(W5500_NODE(inst), reset_gpios),		\
```
`w5500_soft_reset` does the soft reset by resetting the internal register, which IMO will clear the MAC address back to 0. or is it not the case?
From datasheet:
```
If this bit is ‘1’, All internal registers will be initialized. It will be
automatically cleared as ‘0’ after S/W reset.
```

gramsay0 replied:
Yes, a soft reset clears the MAC address.
It is saved after hard reset (in `wiz550io_init`) and written back after soft reset (in `w5500_init`)

parthitce commented on boards/shields/wiznet_wiz550io/wiznet_wiz550io.overlay:
```diff
@@ -0,0 +1,25 @@
+		/* Note: Reducing spi-max-frequency may help with errors/stability */
+		spi-max-frequency = <DT_FREQ_M(80)>;
+		int-gpios = <&arduino_header 15 GPIO_ACTIVE_LOW>;    /* D9 */
+		/* Do not put the reset-gpios property here */
```
Isn't that implicit and specific to this shield. Do we need to mention?

parthitce commented on boards/shields/wiznet_wiz550io/wiznet_wiz550io.overlay:
```diff
@@ -0,0 +1,25 @@
+	eth_w5500: eth_w5500@0 {
+		compatible = "wiznet,w5500";
+		reg = <0x0>;
+		/* Note: Reducing spi-max-frequency may help with errors/stability */
```
33.3 is better as far as datasheet, we can keep 80 without this comment?

parthitce reviewed (commented):
Thanks for the effort in moving to shield. So the random-mac address (if added in the dts) will be used even if unique MAC is available?

gmarull commented on boards/shields/wiznet_wiz550io/CMakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+
+zephyr_library()
+
+# Allow access to private W5500 Ethernet header
+zephyr_library_include_directories(${ZEPHYR_BASE}/drivers/ethernet)
```
looks like a hack, needs a better solution

decsny replied:
do you have an idea?

decsny replied:
@gramsay0 I proposed some ideas in other comments to help address this comment

gmarull commented on boards/shields/wiznet_wiz550io/dts/bindings/wiznet,wiz550io.yaml:
```diff
@@ -0,0 +1,36 @@
+# Copyright (c) 2024 Grant Ramsay
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+  WIZnet WIZ550io - W5500 Ethernet controller.
+  WIZ550io contains a W5500 Ethernet controller with an extra PIC12F519 MCU.
+  The PIC12F519 MCU initialises the W5500 with a unique MAC address after a
+  GPIO reset. The unique MAC address is used if neither
+  "zephyr,random-mac-address" or "local-mac-address" properties are
+  configured in the corresponding W5500 device.
+
+compatible: "wiznet,wiz550io"
+
+include: base.yaml
+
+properties:
+  wiznet,w5500:
+    type: phandle
+    required: true
+    description: The corresponding W5500 device within the WIZ550io module
+  reset-gpios:
+    type: phandle-array
+    required: true
+    description: |
+      Reset pin (active low output).
+      The reset-gpios property must be assigned to WIZ550io rather than
+      the W5500 device. The hardware reset pin is required for WIZ550io,
+      as a software reset of W5500 will clear the MAC address assigned
+      by the PIC12F519.
+  ready-gpios:
+    type: phandle-array
```
please, no shield-level compatibles. This doesn't make any sense.

gmarull commented on boards/shields/wiznet_wiz550io/wiznet_wiz550io.c:
```diff
@@ -0,0 +1,136 @@
+ */
+#define WIZ550IO_RESET_DELAY K_MSEC(150)
+
+static int wiz550io_init(const struct wiz550io_config *config)
```
this could be added as a quirk to the original driver

pdgendt replied:
This has been discussed in both the dev-review and arch working group and has been agreed upon to have this as an exception for a shield, rather than a driver.

gmarull replied:
are meeting replacing reviews now? I strongly disagree in this case.

pdgendt replied:
> are meetings replacing reviews now? 

No, and I never claimed otherwise. I merely stated it, the whole reason #75310 is merged, is to allow these exceptions.

gmarull replied:
I think this can be solved properly without recurring to shield-level code. The shield-level binding alone just doesn't make any sense.

decsny replied:
@gmarull is it the existence of the binding itself that doesnt make sense or the location of it?

decsny replied:
> I think this can be solved properly without recurring to shield-level code

I think it could be solved without shield level code, by handling it in the application, but, this shield code has a benefit of the application not having to handle initialization of the module by providing this reusable code here. I don't see how it is really any different from zephyr handling initialization of many other things in various drivers and platform files. You could write a whole application without using Zephyr, but then what are we doing here?

gramsay0 commented:
> So the random-mac address (if added in the dts) will be used even if unique MAC is available?

@parthitce yes, although setting "local-mac-address" from the device tree was not working.
I have updated with a fix

decsny commented on boards/shields/wiznet_wiz550io/wiznet_wiz550io.c:
```diff
@@ -0,0 +1,143 @@
+	const struct w5500_config *w5500_config = config->w5500_dev->config;
+	struct w5500_runtime *w5500_runtime = config->w5500_dev->data;
+
+	if (!spi_is_ready_dt(&w5500_config->spi)) {
```
the spi could be figured out from the DT in this file instead of looking at the ethernet mac config

decsny commented on boards/shields/wiznet_wiz550io/wiznet_wiz550io.c:
```diff
@@ -0,0 +1,143 @@
+		 * and before software reset.
+		 */
+		err = w5500_spi_read(config->w5500_dev, W5500_SHAR,
+				     w5500_runtime->mac_addr,
```
instead of using the driver runtime from the ethernet controller, I believe that there is an API that can be called to set the mac address https://docs.zephyrproject.org/latest/connectivity/networking/api/ethernet.html#c.ethernet_api.set_config

brandon-exact commented:
How does the application find the `CMakeLists.txt`?

decsny commented:
@gramsay0 I can imagine this PR has been frustrating, but do you plan to return to it?

gramsay0 commented:
> @gramsay0 I can imagine this PR has been frustrating, but do you plan to return to it?

@decsny yes, this PR has been a bit frustrating.

> So my suggestion is this... I think you could basically keep the idea of the C code the same but just add a boards/ folder within the shield directory (see other shields for examples) to allow board specific overlays to add the GPIOs, which could be specified as GPIO hogs (which are children nodes of the gpio controller node, see the gpio controller bindings for details). The gpio hog nodes could be aliased to a standard alias name in the main shield overlay so that the C code can reference it. This way, you dont need a binding for the shield. Similar approach for the reference to the ethernet controller, you could do this with an alias instead of a DT property for a new binding.

This seems like a bit of a hack, a "DT compatible" seems much cleaner and follows the DT/driver pattern used elsewhere.
How would you write a multi-instance driver like this?

> the spi could be figured out from the DT in this file instead of looking at the ethernet mac config

> instead of using the driver runtime from the ethernet controller, I believe that there is an API that can be called to set the mac address

I agree with these points

decsny commented:
> This seems like a bit of a hack, a "DT compatible" seems much cleaner and follows the DT/driver pattern used elsewhere.
> How would you write a multi-instance driver like this?

Okay, I didn't think of it as a hack (and actually thought it was pretty clean) because I was not considering the shield  .c file to be a "driver".... I guess is it expected for a user to use multiple instances of this same shield ? If it is, then yeah maybe we need to refine the idea or have a different one... or maybe for now it would be easier to have this shield be single instance and hopefully get past the blocks on this PR and get this finally merged in?

gramsay0 commented:
@decsny I had a go at making your suggested changes, please take another look

decsny commented:
@gmarull @parthitce please re-review.

gmarull reviewed (changes requested):
should be a driver quirk

pdgendt commented:
> should be a driver quirk

I disagree, the logic is only needed for the shield, so keep it as close to the shield as possible. Image if there'd exist many different variants of the shield, each with there own quirk. That would be unmaintainable in the driver.

This on the other hand, shows a clear separation of concerns, and is the cleanest solution IMO.

gmarull commented:
> > should be a driver quirk
> 
> I disagree, the logic is only needed for the shield, so keep it as close to the shield as possible. Image if there'd exist many different variants of the shield, each with there own quirk. That would be unmaintainable in the driver.
> 
> This on the other hand, shows a clear separation of concerns, and is the cleanest solution IMO.

If they have made a piece of HW that behaves differently that off-the-shelf W5500, they should in a first place describe it in DT as `vnd,wiz550io`. Then a driver that re-uses WIZ5500 code can be created, just with an init call that does any necessary extra work to make the module work as expected. Any board/shield level code is a hack. Just see: `eth_w5500` doesn't even work without the gpio-hogs being run, and the dependency is not captured everywhere. So my -1 stands.

pdgendt commented on boards/shields/wiznet_wiz550io/boards/nucleo_f756zg.overlay:
```diff
@@ -0,0 +1,26 @@
+ * WARNING: The pin PA7 will conflict on selection of SPI_1 and ETH_STM32_HAL.
+ */
+
+/ {
+	aliases {
+		wiz550io-reset = &wiz550io_reset;
+		wiz550io-ready = &wiz550io_ready;
+	};
+};
```
Actually this file should be removed entirely, and moved to `wiznet_wiz550io.overlay` using the `arduino_gpio` label.

decsny commented:
> > > should be a driver quirk
> > 
> > 
> > I disagree, the logic is only needed for the shield, so keep it as close to the shield as possible. Image if there'd exist many different variants of the shield, each with there own quirk. That would be unmaintainable in the driver.
> > This on the other hand, shows a clear separation of concerns, and is the cleanest solution IMO.
> 
> If they have made a piece of HW that behaves differently that off-the-shelf W5500, they should in a first place describe it in DT as `vnd,wiz550io`. Then a driver that re-uses WIZ5500 code can be created, just with an init call that does any necessary extra work to make the module work as expected. Any board/shield level code is a hack. Just see: `eth_w5500` doesn't even work without the gpio-hogs being run, and the dependency is not captured everywhere. So my -1 stands.

You are showing that you don't even understand the PR, this is not a piece of hardware that "behaves differently than the off the shelf W5500", this is a BREAKOUT BOARD with the main purpose of having a W5500 on it. There's no hack, we discussed multiple times that putting board specific gpio signalling in an ethernet mac driver would be the hack. And ARCH WG and dev review discussed this PR each multiple times and discussed and approved the shield level code idea more broadly, so this should not be a blocker for this PR from one person who doesn't agree, we already moved forward as a project with the idea many months ago.

gmarull commented:
> > > > should be a driver quirk
> > > 
> > > 
> > > I disagree, the logic is only needed for the shield, so keep it as close to the shield as possible. Image if there'd exist many different variants of the shield, each with there own quirk. That would be unmaintainable in the driver.
> > > This on the other hand, shows a clear separation of concerns, and is the cleanest solution IMO.
> > 
> > 
> > If they have made a piece of HW that behaves differently that off-the-shelf W5500, they should in a first place describe it in DT as `vnd,wiz550io`. Then a driver that re-uses WIZ5500 code can be created, just with an init call that does any necessary extra work to make the module work as expected. Any board/shield level code is a hack. Just see: `eth_w5500` doesn't even work without the gpio-hogs being run, and the dependency is not captured everywhere. So my -1 stands.
> 
> You are showing that you don't even understand the PR, this is not a piece of hardware that "behaves differently than the off the shelf W5500", this is a BREAKOUT BOARD with the main purpose of having a W5500 on it. There's no hack, we discussed multiple times that putting board specific gpio signalling in an ethernet mac driver would be the hack. And ARCH WG and dev review discussed this PR each multiple times and discussed and approved the shield level code idea more broadly, so this should not be a blocker for this PR from one person who doesn't agree, we already moved forward as a project with the idea many months ago.

Can you then respond to this simple question: does DT capture dependencies properly? Spoiler: no.

decsny commented:
> > > > > should be a driver quirk
> > > > 
> > > > 
> > > > I disagree, the logic is only needed for the shield, so keep it as close to the shield as possible. Image if there'd exist many different variants of the shield, each with there own quirk. That would be unmaintainable in the driver.
> > > > This on the other hand, shows a clear separation of concerns, and is the cleanest solution IMO.
> > > 
> > > 
> > > If they have made a piece of HW that behaves differently that off-the-shelf W5500, they should in a first place describe it in DT as `vnd,wiz550io`. Then a driver that re-uses WIZ5500 code can be created, just with an init call that does any necessary extra work to make the module work as expected. Any board/shield level code is a hack. Just see: `eth_w5500` doesn't even work without the gpio-hogs being run, and the dependency is not captured everywhere. So my -1 stands.
> > 
> > 
> > You are showing that you don't even understand the PR, this is not a piece of hardware that "behaves differently than the off the shelf W5500", this is a BREAKOUT BOARD with the main purpose of having a W5500 on it. There's no hack, we discussed multiple times that putting board specific gpio signalling in an ethernet mac driver would be the hack. And ARCH WG and dev review discussed this PR each multiple times and discussed and approved the shield level code idea more broadly, so this should not be a blocker for this PR from one person who doesn't agree, we already moved forward as a project with the idea many months ago.
> 
> Can you then respond to this simple question: does DT capture dependencies properly? Spoiler: no.

Which dependencies are you talking about? I think the PR as it stands right now clearly shows the DT capturing the situation. There is a w5500 ethernet mac and some gpios that need to be toggled. It's not that complicated.

decsny commented:
> , they should in a first place describe it in DT as `vnd,wiz550io`.

Also, the author had this originally, and in the most recent revision they removed it because of your block.

gmarull commented:
> > > > > > should be a driver quirk
> > > > > 
> > > > > 
> > > > > I disagree, the logic is only needed for the shield, so keep it as close to the shield as possible. Image if there'd exist many different variants of the shield, each with there own quirk. That would be unmaintainable in the driver.
> > > > > This on the other hand, shows a clear separation of concerns, and is the cleanest solution IMO.
> > > > 
> > > > 
> > > > If they have made a piece of HW that behaves differently that off-the-shelf W5500, they should in a first place describe it in DT as `vnd,wiz550io`. Then a driver that re-uses WIZ5500 code can be created, just with an init call that does any necessary extra work to make the module work as expected. Any board/shield level code is a hack. Just see: `eth_w5500` doesn't even work without the gpio-hogs being run, and the dependency is not captured everywhere. So my -1 stands.
> > > 
> > > 
> > > You are showing that you don't even understand the PR, this is not a piece of hardware that "behaves differently than the off the shelf W5500", this is a BREAKOUT BOARD with the main purpose of having a W5500 on it. There's no hack, we discussed multiple times that putting board specific gpio signalling in an ethernet mac driver would be the hack. And ARCH WG and dev review discussed this PR each multiple times and discussed and approved the shield level code idea more broadly, so this should not be a blocker for this PR from one person who doesn't agree, we already moved forward as a project with the idea many months ago.
> > 
> > 
> > Can you then respond to this simple question: does DT capture dependencies properly? Spoiler: no.
> 
> Which dependencies are you talking about? I think the PR as it stands right now clearly shows the DT capturing the situation. There is a w5500 ethernet mac and some gpios that need to be toggled. It's not that complicated.

```

	wiz550io_reset: wiz550io-reset {
		gpio-hog;
		gpios = <12 GPIO_ACTIVE_LOW>; /* D8 */
		output-high;
	};

	wiz550io_ready: wiz550io-ready {
		gpio-hog;
		gpios = <13 GPIO_ACTIVE_HIGH>; /* D7 */
		input;
	};
```

^ There's nothing that indicates these are needed by WIZ5500 to operate correctly. Just C code with SYS_INIT hackery. This simply needs a better solution, and my firm -1 stands.

decsny commented:
> > > > > > > should be a driver quirk
> > > > > > 
> > > > > > 
> > > > > > I disagree, the logic is only needed for the shield, so keep it as close to the shield as possible. Image if there'd exist many different variants of the shield, each with there own quirk. That would be unmaintainable in the driver.
> > > > > > This on the other hand, shows a clear separation of concerns, and is the cleanest solution IMO.
> > > > > 
> > > > > 
> > > > > If they have made a piece of HW that behaves differently that off-the-shelf W5500, they should in a first place describe it in DT as `vnd,wiz550io`. Then a driver that re-uses WIZ5500 code can be created, just with an init call that does any necessary extra work to make the module work as expected. Any board/shield level code is a hack. Just see: `eth_w5500` doesn't even work without the gpio-hogs being run, and the dependency is not captured everywhere. So my -1 stands.
> > > > 
> > > > 
> > > > You are showing that you don't even understand the PR, this is not a piece of hardware that "behaves differently than the off the shelf W5500", this is a BREAKOUT BOARD with the main purpose of having a W5500 on it. There's no hack, we discussed multiple times that putting board specific gpio signalling in an ethernet mac driver would be the hack. And ARCH WG and dev review discussed this PR each multiple times and discussed and approved the shield level code idea more broadly, so this should not be a blocker for this PR from one person who doesn't agree, we already moved forward as a project with the idea many months ago.
> > > 
> > > 
> > > Can you then respond to this simple question: does DT capture dependencies properly? Spoiler: no.
> > 
> > 
> > Which dependencies are you talking about? I think the PR as it stands right now clearly shows the DT capturing the situation. There is a w5500 ethernet mac and some gpios that need to be toggled. It's not that complicated.
> 
> ```
> 
> 	wiz550io_reset: wiz550io-reset {
> 		gpio-hog;
> 		gpios = <12 GPIO_ACTIVE_LOW>; /* D8 */
> 		output-high;
> 	};
> 
> 	wiz550io_ready: wiz550io-ready {
> 		gpio-hog;
> 		gpios = <13 GPIO_ACTIVE_HIGH>; /* D7 */
> 		input;
> 	};
> ```
> 
> ^ There's nothing that indicates these are needed by WIZ5500 to operate correctly. Just C code with SYS_INIT hackery. This simply needs a better solution, and my firm -1 stands.

You're absolutely right, they are not needed specifically by the W5500 mac, which is why we don't want to put it in the driver for the W5500.

gmarull commented:
> > > > > > > > should be a driver quirk
> > > > > > > 
> > > > > > > 
> > > > > > > I disagree, the logic is only needed for the shield, so keep it as close to the shield as possible. Image if there'd exist many different variants of the shield, each with there own quirk. That would be unmaintainable in the driver.
> > > > > > > This on the other hand, shows a clear separation of concerns, and is the cleanest solution IMO.
> > > > > > 
> > > > > > 
> > > > > > If they have made a piece of HW that behaves differently that off-the-shelf W5500, they should in a first place describe it in DT as `vnd,wiz550io`. Then a driver that re-uses WIZ5500 code can be created, just with an init call that does any necessary extra work to make the module work as expected. Any board/shield level code is a hack. Just see: `eth_w5500` doesn't even work without the gpio-hogs being run, and the dependency is not captured everywhere. So my -1 stands.
> > > > > 
> > > > > 
> > > > > You are showing that you don't even understand the PR, this is not a piece of hardware that "behaves differently than the off the shelf W5500", this is a BREAKOUT BOARD with the main purpose of having a W5500 on it. There's no hack, we discussed multiple times that putting board specific gpio signalling in an ethernet mac driver would be the hack. And ARCH WG and dev review discussed this PR each multiple times and discussed and approved the shield level code idea more broadly, so this should not be a blocker for this PR from one person who doesn't agree, we already moved forward as a project with the idea many months ago.
> > > > 
> > > > 
> > > > Can you then respond to this simple question: does DT capture dependencies properly? Spoiler: no.
> > > 
> > > 
> > > Which dependencies are you talking about? I think the PR as it stands right now clearly shows the DT capturing the situation. There is a w5500 ethernet mac and some gpios that need to be toggled. It's not that complicated.
> > 
> > 
> > ```
> > 
> > 	wiz550io_reset: wiz550io-reset {
> > 		gpio-hog;
> > 		gpios = <12 GPIO_ACTIVE_LOW>; /* D8 */
> > 		output-high;
> > 	};
> > 
> > 	wiz550io_ready: wiz550io-ready {
> > 		gpio-hog;
> > 		gpios = <13 GPIO_ACTIVE_HIGH>; /* D7 */
> > 		input;
> > 	};
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ^ There's nothing that indicates these are needed by WIZ5500 to operate correctly. Just C code with SYS_INIT hackery. This simply needs a better solution, and my firm -1 stands.
> 
> You're absolutely right, they are not needed specifically by the W5500 mac, which is why we don't want to put it in the driver for the W5500.

Then please, describe this properly in DT at the very least. Then request a re-review so I can provide feedback.

decsny commented:
> > > > > > > > > should be a driver quirk
> > > > > > > > 
> > > > > > > > 
> > > > > > > > I disagree, the logic is only needed for the shield, so keep it as close to the shield as possible. Image if there'd exist many different variants of the shield, each with there own quirk. That would be unmaintainable in the driver.
> > > > > > > > This on the other hand, shows a clear separation of concerns, and is the cleanest solution IMO.
> > > > > > > 
> > > > > > > 
> > > > > > > If they have made a piece of HW that behaves differently that off-the-shelf W5500, they should in a first place describe it in DT as `vnd,wiz550io`. Then a driver that re-uses WIZ5500 code can be created, just with an init call that does any necessary extra work to make the module work as expected. Any board/shield level code is a hack. Just see: `eth_w5500` doesn't even work without the gpio-hogs being run, and the dependency is not captured everywhere. So my -1 stands.
> > > > > > 
> > > > > > 
> > > > > > You are showing that you don't even understand the PR, this is not a piece of hardware that "behaves differently than the off the shelf W5500", this is a BREAKOUT BOARD with the main purpose of having a W5500 on it. There's no hack, we discussed multiple times that putting board specific gpio signalling in an ethernet mac driver would be the hack. And ARCH WG and dev review discussed this PR each multiple times and discussed and approved the shield level code idea more broadly, so this should not be a blocker for this PR from one person who doesn't agree, we already moved forward as a project with the idea many months ago.
> > > > > 
> > > > > 
> > > > > Can you then respond to this simple question: does DT capture dependencies properly? Spoiler: no.
> > > > 
> > > > 
> > > > Which dependencies are you talking about? I think the PR as it stands right now clearly shows the DT capturing the situation. There is a w5500 ethernet mac and some gpios that need to be toggled. It's not that complicated.
> > > 
> > > 
> > > ```
> > > 
> > > 	wiz550io_reset: wiz550io-reset {
> > > 		gpio-hog;
> > > 		gpios = <12 GPIO_ACTIVE_LOW>; /* D8 */
> > > 		output-high;
> > > 	};
> > > 
> > > 	wiz550io_ready: wiz550io-ready {
> > > 		gpio-hog;
> > > 		gpios = <13 GPIO_ACTIVE_HIGH>; /* D7 */
> > > 		input;
> > > 	};
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > ^ There's nothing that indicates these are needed by WIZ5500 to operate correctly. Just C code with SYS_INIT hackery. This simply needs a better solution, and my firm -1 stands.
> > 
> > 
> > You're absolutely right, they are not needed specifically by the W5500 mac, which is why we don't want to put it in the driver for the W5500.
> 
> Then please, describe this properly in DT at the very least. Then request a re-review so I can provide feedback.

What would make it "more proper" in your opinion

nashif commented:
![image](https://github.com/user-attachments/assets/c23f3ed1-b113-4700-aac9-592be64c7069)

please follow the escalation process described here: https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#pr-technical-escalation

decsny commented:
> ![image](https://private-user-images.githubusercontent.com/180017/374997059-c23f3ed1-b113-4700-aac9-592be64c7069.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg0ODY5NjcsIm5iZiI6MTcyODQ4NjY2NywicGF0aCI6Ii8xODAwMTcvMzc0OTk3MDU5LWMyM2YzZWQxLWIxMTMtNDcwMC1hYWM5LTU5MmJlNjRjNzA2OS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDA5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAwOVQxNTExMDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kZDhjYzg3YTI2MGU3MDE1MmIwYWRiYjlmMTVkN2M5OWE0NmEzNGIwNzY3ODk1ZTVkNzlhZGUzYWM2MTRkMDFhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.vbr7lM0pXOyJ9QfwTPOr9F8vuOB4HFhjQNwNdIMJ6n8)
> 
> please follow the escalation process described here: https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#pr-technical-escalation

We have, this PR has been discussed in the Arch WG and dev review meeting multiple times, and the stated change request is in opposition to the conclusions agreed in those meetings

gmarull commented:
> > > > > > > > > > should be a driver quirk
> > > > > > > > > 
> > > > > > > > > 
> > > > > > > > > I disagree, the logic is only needed for the shield, so keep it as close to the shield as possible. Image if there'd exist many different variants of the shield, each with there own quirk. That would be unmaintainable in the driver.
> > > > > > > > > This on the other hand, shows a clear separation of concerns, and is the cleanest solution IMO.
> > > > > > > > 
> > > > > > > > 
> > > > > > > > If they have made a piece of HW that behaves differently that off-the-shelf W5500, they should in a first place describe it in DT as `vnd,wiz550io`. Then a driver that re-uses WIZ5500 code can be created, just with an init call that does any necessary extra work to make the module work as expected. Any board/shield level code is a hack. Just see: `eth_w5500` doesn't even work without the gpio-hogs being run, and the dependency is not captured everywhere. So my -1 stands.
> > > > > > > 
> > > > > > > 
> > > > > > > You are showing that you don't even understand the PR, this is not a piece of hardware that "behaves differently than the off the shelf W5500", this is a BREAKOUT BOARD with the main purpose of having a W5500 on it. There's no hack, we discussed multiple times that putting board specific gpio signalling in an ethernet mac driver would be the hack. And ARCH WG and dev review discussed this PR each multiple times and discussed and approved the shield level code idea more broadly, so this should not be a blocker for this PR from one person who doesn't agree, we already moved forward as a project with the idea many months ago.
> > > > > > 
> > > > > > 
> > > > > > Can you then respond to this simple question: does DT capture dependencies properly? Spoiler: no.
> > > > > 
> > > > > 
> > > > > Which dependencies are you talking about? I think the PR as it stands right now clearly shows the DT capturing the situation. There is a w5500 ethernet mac and some gpios that need to be toggled. It's not that complicated.
> > > > 
> > > > 
> > > > ```
> > > > 
> > > > 	wiz550io_reset: wiz550io-reset {
> > > > 		gpio-hog;
> > > > 		gpios = <12 GPIO_ACTIVE_LOW>; /* D8 */
> > > > 		output-high;
> > > > 	};
> > > > 
> > > > 	wiz550io_ready: wiz550io-ready {
> > > > 		gpio-hog;
> > > > 		gpios = <13 GPIO_ACTIVE_HIGH>; /* D7 */
> > > > 		input;
> > > > 	};
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > ^ There's nothing that indicates these are needed by WIZ5500 to operate correctly. Just C code with SYS_INIT hackery. This simply needs a better solution, and my firm -1 stands.
> > > 
> > > 
> > > You're absolutely right, they are not needed specifically by the W5500 mac, which is why we don't want to put it in the driver for the W5500.
> > 
> > 
> > Then please, describe this properly in DT at the very least. Then request a re-review so I can provide feedback.
> 
> What would make it "more proper" in your opinion

Any solution that represents dependencies in DT correctly, ie, WIZ550io GPIO ports ordinal < WIZ5500 in all cases. Also, it looks like this solution is accessing internals:

```
		const struct ethernet_api *w5500_api = w5500_dev->api;
		struct ethernet_config config = {.mac_address = wiz550io_mac_addr};

		ret = w5500_api->set_config(w5500_dev, ETHERNET_CONFIG_TYPE_MAC_ADDRESS, &config);
```

Overall, this PR is a dirty hack.

gmarull commented:
> > ![image](https://private-user-images.githubusercontent.com/180017/374997059-c23f3ed1-b113-4700-aac9-592be64c7069.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg0ODY5NjcsIm5iZiI6MTcyODQ4NjY2NywicGF0aCI6Ii8xODAwMTcvMzc0OTk3MDU5LWMyM2YzZWQxLWIxMTMtNDcwMC1hYWM5LTU5MmJlNjRjNzA2OS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDA5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAwOVQxNTExMDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kZDhjYzg3YTI2MGU3MDE1MmIwYWRiYjlmMTVkN2M5OWE0NmEzNGIwNzY3ODk1ZTVkNzlhZGUzYWM2MTRkMDFhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.vbr7lM0pXOyJ9QfwTPOr9F8vuOB4HFhjQNwNdIMJ6n8)
> > please follow the escalation process described here: https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#pr-technical-escalation
> 
> We have, this PR has been discussed in the Arch WG and dev review meeting multiple times, and the stated change request is in opposition to the conclusions agreed in those meetings

Meetings to not replace reviews, have I missed something?

bjarki-andreasen commented on boards/shields/wiznet_wiz550io/boards/nucleo_f756zg.overlay:
```diff
@@ -0,0 +1,26 @@
+&gpiof {
+	status = "okay";
+
+	wiz550io_reset: wiz550io-reset {
+		gpio-hog;
+		gpios = <12 GPIO_ACTIVE_LOW>; /* D8 */
+		output-high;
+	};
+
+	wiz550io_ready: wiz550io-ready {
+		gpio-hog;
+		gpios = <13 GPIO_ACTIVE_HIGH>; /* D7 */
+		input;
+	};
```
hogs are only used for pins which are not configured by the system, both reset and ready are initialized by the driver or shield code, they shall be initialized from there :)

bjarki-andreasen commented on boards/shields/wiznet_wiz550io/wiznet_wiz550io.c:
```diff
@@ -0,0 +1,157 @@
+	return 0;
+}
+
+static int wiz550io_post_init(void)
+{
+	int ret = 0;
+
+	if (WIZ550IO_USE_MAC_ADDRESS) {
+		/* Assign the WIZ550io unique MAC address */
+		const struct device *w5500_dev = DEVICE_DT_GET(W5500_NODE);
+		const struct ethernet_api *w5500_api = w5500_dev->api;
+		struct ethernet_config config = {.mac_address = wiz550io_mac_addr};
+
+		ret = w5500_api->set_config(w5500_dev, ETHERNET_CONFIG_TYPE_MAC_ADDRESS, &config);
+	}
+
+	return ret;
+}
```
This should be moved to the w5500 device driver, implemented in the `get_config()` function of the `ethernet_api`.

This code is essentially an ad-hoc implementation of part of the `get_config()` API, which is then used to overwrite the config using `set_config()`

bjarki-andreasen replied:
See
https://github.com/zephyrproject-rtos/zephyr/blob/9b5260de9a31738f5632a0f7240f1f70bc1d4cd4/drivers/ethernet/eth_w5500.c#L462-L469
and
https://github.com/zephyrproject-rtos/zephyr/blob/9b5260de9a31738f5632a0f7240f1f70bc1d4cd4/include/zephyr/net/ethernet.h#L557-L560

gramsay0 replied:
@bjarki-andreasen I'm not sure I understand, could you elaborate?

It looks like `get_config` is used by `net_eth_get_hw_config` to "Return ethernet device hardware configuration information". How does this help to set the mac address?

decsny replied:
Strictly speaking, you could argue this breaks encapsulation, but I think it's fine since this board is basically just the w5500, there's not much to be gained from encapsulation. And chucking board specific code into the ethernet mac driver any time somebody introduces a breakout board with this mac doesn't make any sense. That would quickly get out of hand. 

I have absolutely no idea why so many people are complaining about breaking the abstraction of the ethernet api by calling a driver function from the (known) device struct while simultaneously suggesting to put a bunch of board specific code in the MAC driver, which is an even worse breach of abstraction, IMO. Can somebody explain to me the thought process there instead of just saying to do these things, we already talked a lot about this on this PR, and let's stop going back and forth on things.

bjarki-andreasen replied:
The MAC address is read from the WF5500 chip AFAIK, therefore this is not violating any encapsulation. All the external chip does is set the MAC address using the same interface as `set_config()` does. No board specific code is going into the MAC driver, but MAC specific code is moved to the MAC driver from the board code.

bjarki-andreasen replied:
@gramsay0 The MAC address is one of those hardware configurations :) So when a higher level subsystem "probes" the device using `get_config()`, it will retrieve and set the MAC address which has been assigned to the WF5500 by the external chip.

gramsay0 replied:
@bjarki-andreasen ahh thanks I understand now.

The problem is reset sequencing. A hard (GPIO) reset is required for the external chip to write the MAC address into W5500. But then the soft (SPI command) reset in W5500 init will clear it.

Earlier we discussed disabling the soft reset in W5500 init, although it's unclear exactly which registers the external chip modifes. It seemed safer to keep the soft reset to put the device in a known state, the unique MAC is all we are interested in

nordicjm replied:
You do care about the MAC address because it is an OUI number for which you have to buy an identifier and have a block of addresses to assign to devices (with one unique number per device). You can't just give devices random addresses, despite some manufacturers being nonchalant on the matter, it is not legal

bjarki-andreasen commented on boards/shields/wiznet_wiz550io/wiznet_wiz550io.overlay:
```diff
@@ -0,0 +1,17 @@
+/* Copyright (c) 2024 Grant Ramsay
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
```
add the following here:
```
/ {
        zephyr,user {
                wiz550io-reset-gpios = <&arduino_header 14 GPIO_ACTIVE_HIGH>;
                wiz550io-ready-gpios = <&arduino_header 13 GPIO_ACTIVE_HIGH>;
        };
};
```
and get them from the shield code like
```
const struct gpio_dt_spec reset_pin = GPIO_DT_SPEC_GET(DT_PATH(zephyr_user), wiz550io_reset_gpios);
```

gramsay0 replied:
I wasn't aware of `zephyr,user` node, I prefer it to GPIO hogs.

Although, looking at the documentation, I doubt it would be allowed:

> This node is meant for sample code and user applications. It should not be used in the upstream Zephyr source code for device drivers, subsystems, etc.

bjarki-andreasen replied:
Using a binding for the board seems appropriate :) Please avoid introducing the `wiz550io-` part of the gpios properties for the board, that was only required to avoid name collisions in the `zephyr,user` node.

bjarki-andreasen reviewed (changes requested):
With the comments I have added we should be able to solve this thing :)

raveious commented on boards/shields/wiznet_wiz550io/wiznet_wiz550io.overlay:
```diff
@@ -0,0 +1,26 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+&arduino_spi {
```
It looks like the Arduino intermediate board that converts the connector on this board to the Arduino pinout is discontinued. Not sure if we care about the inability to buy a required piece of this or not.

In general though, since this shield isn't using the Arduino-style connector (e.g. pin locations and physical dimensions of the connectors), should this be binding directly to the arduino_spi alias?
